src/classes/Collection.cls

```diff
@@ -2,317 +2,494 @@ public with sharing class Collection {
 
 	private List<SObject> records;
 
+
+	// CONSTRUCT
+
 	public static Collection of(Iterable<SObject> records) {
 		return new Collection(records);
 	}
 
+
 	public static Collection of(Set<SObject> records) {
 		return new Collection(new List<SObject>(records));
 	}
 
+
 	private Collection(Iterable<SObject> records) {
 		this.records = new List<SObject>();
+
 		Iterator<SObject> iter = records.iterator();
-		while (iter.hasNext()) {
+
+		while(iter.hasNext()) {
 			this.records.add(iter.next());
 		}
 	}
 
+
+	// CONVERT
+
 	public List<SObject> asList() {
 		return new List<SObject>(records);
 	}
 
+
 	public List<SObject> asList(Type listType) {
-		List<SObject> typedList = (List<SObject>) listType.newInstance();
+		List<SObject> typedList =(List<SObject>) listType.newInstance();
 		typedList.addAll(records);
 		return typedList;
 	}
 
+
 	public Set<SObject> asSet() {
 		return new Set<SObject>(records);
 	}
 
+
 	public Set<SObject> asSet(Type setType) {
-		Set<SObject> typedSet = (Set<SObject>) setType.newInstance();
+		Set<SObject> typedSet =(Set<SObject>) setType.newInstance();
 		typedSet.addAll(records);
 		return typedSet;
 	}
 
+
+	public Map<Id, SObject> asMap() {
+		return new Map<Id, SObject>(records);
+	}
+
+
+	public Map<Id, SObject> asMap(Type mapType) {
+		Map<Id, SObject> typedMap =(Map<Id, SObject>) mapType.newInstance();
+		typedMap.putAll(records);
+		return typedMap;
+	}
+
+
+	// FILTER
+
 	public Collection filter(SObjectPredicate predicate) {
-		List<SObject> filtered = new List<SObject>();
-		for (SObject record : records) {
-			if (predicate.apply(record)) {
-				filtered.add(record);
+		List<SObject> result = new List<SObject>();
```

Comment:
`result` as opposed to `filtered` is less clear to what is being done, so this change is not required.